Added a compile time assertion that forbids classes that offer WeakPtr

geoffreygaren opened the pull request:
#### c13321a96e1de0debe0e734a4b092001b22854aa
<pre>
Added a compile time assertion that forbids classes that offer WeakPtr
with no corresponding RefPtr or CheckedPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=273569">https://bugs.webkit.org/show_bug.cgi?id=273569</a>
<a href="https://rdar.apple.com/127296969">rdar://127296969</a>

Reviewed by Ryosuke Niwa.

An object that offers WeakPtr with no corresponding RefPtr or CheckedPtr
is a dangerous contradiction. On the one hand, we know that it can be
deleted at any time. On the other hand, when we go to use it, we do
nothing to ensure its lifetime. This has been a source of use after
free / security bugs.

This patch adds a compile time assertion against future uses of this
anti-pattern. An explicit template specialization allow list maintains
existing uses. The allow list is a todo list for future deployment of
RefPtr or CheckedPtr.

This patch does not enforce correct usage of RefPtr or CheckedPtr; it
only enforces a rule that a class must at least offer RefPtr or
CheckedPtr if it offers WeakPtr.

In order to facilitate the allow list, I needed to
    * Move some nested classes out-of-line, since C++ does not support
    forward declaration or template specialization of nested classes

    * Move some destructors out-of-line, since our compile time
    assertion requires a complete type definition, and I didn&apos;t want to
    increase #includes in headers

Canonical link: <a href="https://commits.webkit.org/278224@main">https://commits.webkit.org/278224@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96b57700305c3318eaae5117a6689dc017fdbdc2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2189 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/53151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52209 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/35260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/53151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52006 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/35260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/53151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/35260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-skia~~](https://ews-build.webkit.org/#/builders/52/builds/8278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/43230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/35260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/54732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49403 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/25002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/54732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/43132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/54732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/27121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/56886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7194 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/56886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/TypeTraits.h:
```diff
@@ -92,25 +92,40 @@ struct RemoveSmartPointerHelper<T, Ref<Pointee>> {
 
 template<class T>
-static auto HasRefCountMethodsTest(int) -> SFINAE1True<decltype(std::declval<T>().ref(), std::declval<T>().deref())>;
+static auto HasRefPtrMethodsTest(int) -> SFINAE1True<decltype(&T::ref, &T::deref)>;
```
Do we really need int argument here?

geoffreygaren replied:
The int vs long argument overload is one of the -- admittedly very weird -- customary ways to do C++ SFINAE programming. 'int' is preferred over 'long' in overload resolution, so the compiler will select the 'int' version if it is not an error, and otherwise select the 'long' version. So, in a very weird way, 'int' means "use this if possible" and 'long' means "otherwise default to this if necessary".

rniwa replied:
oh, I see. That's pretty counterintuitive. Maybe we should add a comment there explaining as such. Or maybe in the commit message?

rniwa commented on Source/WTF/wtf/TypeTraits.h:
```diff
@@ -92,25 +92,40 @@ struct RemoveSmartPointerHelper<T, Ref<Pointee>> {
+static auto HasRefPtrMethodsTest(int) -> SFINAE1True<decltype(&T::ref, &T::deref)>;
 template<class>
-static auto HasRefCountMethodsTest(long) -> std::false_type;
+static auto HasRefPtrMethodsTest(long) -> std::false_type;
```
Ditto for long.

rniwa commented on Source/WTF/wtf/WeakRef.h:
```diff
@@ -31,9 +31,84 @@
 #include <wtf/WeakPtrImpl.h>
 
 namespace WTF {
+    template<typename Value, typename HashFunctions> class ListHashSet;
```
I think we don't normally indent thing inside `namespace`.

geoffreygaren replied:
Will fix once I get it building.

rniwa commented on Source/WebCore/html/shadow/DateTimeFieldElement.h:
```diff
@@ -33,34 +33,44 @@
 
 enum class DateTimePlaceholderIfNoValue : bool { No, Yes };
 
+class FieldOwner : public CanMakeWeakPtr<FieldOwner> {
```
Maybe this should be `DateTimeFieldElementFieldOwner`?

rniwa reviewed (approved):
r=me once EWS turns green.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -106,6 +106,15 @@ OBJC_CLASS NSEvent;
+    template<typename T> struct IsDeprecatedWeakRefSmartPointerException;
+    template<> struct IsDeprecatedWeakRefSmartPointerException<WebKit::WebExtensionContext> : std::true_type { };
+}
+
```
@geoffreygaren As an API object this has `ref()` and `deref()` from `API::Object` and `WeakPtr` from `IPC::MessageReceiver`. It seems like that should be fine and not need an exception for this class?

This is the same situation for other WebExtension classes in `UIProcess/Extensions`, so I'm not sure why this was the only one needing an exception.

Looking beyond extensions, this is the same case for other API objects like `WebPageProxy`, etc.

geoffreygaren replied:
I'll investigate. One weakness to my approach is that the compilation failure can be triggered by forgetting an #include. I had like 500 to get through, so I didn't pay close attention. Maybe I can fix these with an #include.

geoffreygaren replied:
I think it was just this one. Fixed.

geoffreygaren committed:
Added a compile time assertion that forbids classes that offer WeakPtr
with no corresponding RefPtr or CheckedPtr
https://bugs.webkit.org/show_bug.cgi?id=273569
rdar://127296969

Reviewed by Ryosuke Niwa.

An object that offers WeakPtr with no corresponding RefPtr or CheckedPtr
is a dangerous contradiction. On the one hand, we know that it can be
deleted at any time. On the other hand, when we go to use it, we do
nothing to ensure its lifetime. This has been a source of use after
free / security bugs.

This patch adds a compile time assertion against future uses of this
anti-pattern. An explicit template specialization allow list maintains
existing uses. The allow list is a todo list for future deployment of
RefPtr or CheckedPtr.

This patch does not enforce correct usage of RefPtr or CheckedPtr; it
only enforces a rule that a class must at least offer RefPtr or
CheckedPtr if it offers WeakPtr.

In order to facilitate the allow list, I needed to
    * Move some nested classes out-of-line, since C++ does not support
    forward declaration or template specialization of nested classes

    * Move some destructors out-of-line, since our compile time
    assertion requires a complete type definition, and I didn't want to
    increase #includes in headers

Canonical link: https://commits.webkit.org/278224@main
